Reworked Fire Types

caelwarner opened the pull request:
- Before there was the possibility for RenderMaterial to be loaded on dedicated servers, crashing them. With this small rework that can no longer happen.
- Also switched to use ResourceLocation instead of a simple string. This will allow better compat if multiple mods have the same name for their fire
- I hope to eventually move most of this code to an `api` package where other mods can add their own fire types. We could also move the shroomloin type code to that `api` package.
- Fixes #309 fixes #326

caelwarner committed:
Reworked fire types to make them safer

- Before there was the possibility for RenderMaterial to be loaded on dedicated servers, crashing them. With this small rework that can no longer happen.
- Also switched to use ResourceLocation instead of a simple string. This will allow better compat if multiple mods have the same name for their fire
- Removed ClientFireType

SwanX1 commented on src/main/java/org/infernalstudios/infernalexp/access/FireTypeAccess.java:
```diff
@@ -16,78 +16,81 @@
-        ENDER_FIRE("ender_fire", () -> new ClientFireType(LOCATION_ENDER_FIRE_0, LOCATION_ENDER_FIRE_1)),
-        BORIC_FIRE("boric_fire", () -> new ClientFireType(LOCATION_BORIC_FIRE_0, LOCATION_BORIC_FIRE_1)),
-        CRYPTIC_FIRE("cryptic_fire", () -> new ClientFireType(LOCATION_CRYPTIC_FIRE_0, LOCATION_CRYPTIC_FIRE_1));
+    enum FireTypes {
```
```suggestion
    public static enum FireTypes {
```

caelwarner replied:
This inner `enum` is within an `interface` so adding `public` and `static` is redundant

SwanX1 replied:
right... yet we still use `private Object field;`, no?
it's better to be explicit rather than implicit

caelwarner replied:
Because adding `private` is not redundant. By default, members are package private, not private

SwanX1 commented on src/main/java/org/infernalstudios/infernalexp/access/FireTypeAccess.java:
```diff
@@ -16,78 +16,81 @@
         }
     }
 
+    class FireType {
```
```suggestion
    public static class FireType {
```

caelwarner replied:
Same as above

caelwarner committed:
Reworked fire types again

- Made it easier for other mods to add their own fire types
- Moved FireType to api package
- Added IEFireTypes

LudoCrypt commented on src/main/java/org/infernalstudios/infernalexp/api/FireType.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    @OnlyIn(Dist.CLIENT)
+    public RenderMaterial getSprite0() {
+        return RenderMaterialCache.getOrCreate(this.sprite0);
+    }
+
+    @OnlyIn(Dist.CLIENT)
+    public RenderMaterial getSprite1() {
+        return RenderMaterialCache.getOrCreate(this.sprite1);
+    }
+
+    @OnlyIn(Dist.CLIENT)
+    private static final class RenderMaterialCache {
+
+        private static final Map<ResourceLocation, RenderMaterial> RENDER_MATERIALS = new HashMap<>();
+
+        private static RenderMaterial getOrCreate(ResourceLocation resourceLocation) {
+            if (!RENDER_MATERIALS.containsKey(resourceLocation))
+                RENDER_MATERIALS.put(resourceLocation, new RenderMaterial(AtlasTexture.LOCATION_BLOCKS_TEXTURE, resourceLocation));
+
+            return RENDER_MATERIALS.get(resourceLocation);
+        }
+
+    }
```
Switch these to use a Util.memoize
```java
@OnlyIn(Dist.CLIENT)
private static final Function<ResourceLocation, RenderMaterial> SPRITE_GETTER = Util.memoize((resourceLocation) -> new RenderMaterial(AtlasTexture.LOCATION_BLOCKS_TEXTURE, resourceLocation));
```

LudoCrypt replied:
```java
    @OnlyIn(Dist.CLIENT)
    private static final Function<ResourceLocation, RenderMaterial> SPRITE_GETTER = Util.memoize((resourceLocation) -> new RenderMaterial(AtlasTexture.LOCATION_BLOCKS_TEXTURE, resourceLocation));

    @OnlyIn(Dist.CLIENT)
    public RenderMaterial getSprite0() {
        return SPRITE_GETTER.apply(this.sprite0);
    }

    @OnlyIn(Dist.CLIENT)
    public RenderMaterial getSprite1() {
        return SPRITE_GETTER.apply(this.sprite1);
    }
```

caelwarner replied:
The only issue is that I can't add `@OnlyIn` annotations to fields. They can only be added to classes and methods.

LudoCrypt replied:
really? thats dumb

SwanX1 replied:
> The only issue is that I can't add `@OnlyIn` annotations to fields. They can only be added to classes and methods.

The getter function won't be accessed on the server, so there shouldn't be a problem when you're accessing it through a client-only method. The suppliers won't get executed and there shouldn't be any errors.

caelwarner replied:
What package is `Util#memoize` from?

LudoCrypt replied:
net.minecraft

caelwarner replied:
I don't see that. I wonder if it has a different name with MCP mappings

LudoCrypt replied:
linkie says its net.minecraft.util

caelwarner replied:
But only from 1.17 and up

LudoCrypt replied:
Check what the RenderType class uses, that uses memoize.
